src/main/java/gregtech/common/covers/CoverFluidRegulator.java

```diff
@@ -0,0 +1,217 @@
+
+import gregtech.api.cover.ICoverable;
+import gregtech.api.gui.GuiTextures;
+import gregtech.api.gui.ModularUI;
+import gregtech.api.gui.widgets.*;
+import gregtech.api.util.GTFluidUtils;
+import gregtech.api.util.TextFormattingUtil;
+import net.minecraft.entity.player.EntityPlayer;
+import net.minecraft.nbt.NBTTagCompound;
+import net.minecraft.util.EnumFacing;
+import net.minecraft.util.math.MathHelper;
+import net.minecraft.util.text.ITextComponent;
+import net.minecraft.util.text.TextComponentString;
+import net.minecraft.util.text.TextComponentTranslation;
+import net.minecraft.util.text.event.HoverEvent;
+import net.minecraftforge.fluids.FluidStack;
+import net.minecraftforge.fluids.capability.IFluidHandler;
+import net.minecraftforge.fluids.capability.IFluidTankProperties;
+
+import java.util.List;
+import java.util.function.Predicate;
+
+
+public class CoverFluidRegulator extends CoverPump {
+
+    protected TransferMode transferMode;
+    protected int keepAmount = 0;
+    protected int supplyAmount = 0;
+    private static final String supplyKey = "SupplyAmount";
+    private static final String keepKey = "KeepAmount";
+
+    public CoverFluidRegulator(ICoverable coverHolder, EnumFacing attachedSide, int tier, int mbPerTick) {
+        super(coverHolder, attachedSide, tier, mbPerTick);
+        this.transferMode = TransferMode.TRANSFER_ANY;
+    }
+
+    @Override
+    protected int doTransferFluidsInternal(IFluidHandler myFluidHandler, IFluidHandler fluidHandler, int transferLimit) {
+        IFluidHandler sourceHandler;
+        IFluidHandler destHandler;
+
+        if (pumpMode == PumpMode.IMPORT) {
+            sourceHandler = fluidHandler;
+            destHandler = myFluidHandler;
+        } else if (pumpMode == PumpMode.EXPORT) {
+            sourceHandler = myFluidHandler;
+            destHandler = fluidHandler;
+        } else {
+            return 0;
+        }
+        switch (transferMode) {
+            case TRANSFER_ANY: return GTFluidUtils.transferFluids(sourceHandler, destHandler, transferLimit, fluidFilter::testFluidStack);
+            case KEEP_EXACT: return doKeepExact(transferLimit, sourceHandler, destHandler, fluidFilter::testFluidStack, this.keepAmount);
+            case TRANSFER_EXACT: return doTransferExact(transferLimit, sourceHandler, destHandler, fluidFilter::testFluidStack, this.supplyAmount);
+        }
+        return 0;
+    }
+
+    protected int doTransferExact(int transferLimit, IFluidHandler sourceHandler, IFluidHandler destHandler, Predicate<FluidStack> fluidFilter, int supplyAmount) {
+        int fluidLeftToTransfer = transferLimit;
+        for (IFluidTankProperties tankProperties : sourceHandler.getTankProperties()) {
+            if (fluidLeftToTransfer < supplyAmount)
+                break;
+            FluidStack sourceFluid = tankProperties.getContents();
+            if (sourceFluid == null || sourceFluid.amount == 0 || !fluidFilter.test(sourceFluid)) continue;
```

Comment:
Do we want to add another check here to check if the fluid being transferred matches the one in the destination tank, to short circuit some more logic if the fluids don't match?

Reply:
We could but this is already handled [here](https://github.com/GregTechCE/GregTech/blob/ba57b3f7f7dc548cf0abd8d65bca173a1977868e/src/main/java/gregtech/api/util/GTFluidUtils.java#L56). I used the same logic flow as the Pumps use to remain consistent and they don't do this.